src/lib/x509/x509_ext.cpp

```diff
@@ -1283,48 +1282,25 @@ void IPAddressBlocks::IPAddressOrRange<V>::encode_into(Botan::DER_Encoder& into)
 
 template <IPAddressBlocks::Version V>
 void IPAddressBlocks::IPAddressOrRange<V>::decode_from(Botan::BER_Decoder& from) {
-   const size_t version_octets = static_cast<size_t>(V);
-
    const ASN1_Type next_tag = from.peek_next_object().type_tag();
 
    // this can either be a prefix or a single address
    if(next_tag == ASN1_Type::BitString) {
       // construct a min and a max address from the prefix
 
-      std::vector<uint8_t> prefix;
-      from.decode(prefix, ASN1_Type::OctetString, ASN1_Type::BitString, ASN1_Class::Universal);
-
-      // we have to account for the octet at the beginning that specifies how many bits are unused in the last octet
-      if(prefix.empty() || prefix.size() > version_octets + 1) {
-         throw Decoding_Error(
-            fmt("IP address range entries must have a length between 2 and {} bits.", version_octets));
-      }
-
-      const uint8_t unused = prefix.front();
-      const uint8_t discarded_octets = version_octets - (static_cast<uint8_t>(prefix.size()) - 1);
+      std::vector<uint8_t> prefix_min;
+      from.decode(prefix_min, ASN1_Type::OctetString, ASN1_Type::BitString, ASN1_Class::Universal);
 
-      prefix.erase(prefix.begin());
-
-      if(prefix.empty() && unused != 0) {
-         throw Decoding_Error("IP address range entry specified unused bits, but did not provide any octets.");
-      }
-
-      for(size_t i = 0; i < discarded_octets; i++) {
-         prefix.push_back(0);
-      }
+      // copy because we modify the address in `decode_single_address`, but we need it twice for min and max
+      std::vector<uint8_t> prefix_max(prefix_min);
 
-      m_min = IPAddress<V>(prefix);
+      // min address gets filled with 0's
+      decode_single_address(prefix_min, true);
```

Comment:
Every call to `decode_single_address` immediately converts the vector to an `IPAddress<V>` - why not just return that directly?